lgc/builder/BuilderRecorder.cpp

```diff
@@ -1848,10 +1847,147 @@ Instruction *BuilderRecorder::record(BuilderRecorder::Opcode opcode, Type *resul
+    case ACos:
+    case ACosh:
+    case ASin:
+    case ASinh:
+    case ATan:
+    case ATan2:
+    case ATanh:
+    case Cosh:
+    case Determinant:
+    case Exp:
+    case InverseSqrt:
+    case Log:
+    case MatrixInverse:
+    case Opcode::CrossProduct:
+    case Opcode::CubeFaceCoord:
+    case Opcode::CubeFaceIndex:
+    case Opcode::Derivative:
+    case Opcode::DotProduct:
+    case Opcode::ExtractBitField:
+    case Opcode::ExtractExponent:
+    case Opcode::ExtractSignificand:
+    case Opcode::FClamp:
+    case Opcode::FMax:
+    case Opcode::FMax3:
+    case Opcode::FMid3:
+    case Opcode::FMin:
+    case Opcode::FMin3:
+    case Opcode::FMix:
+    case Opcode::FMod:
+    case Opcode::FSign:
+    case Opcode::FaceForward:
+    case Opcode::FindSMsb:
+    case Opcode::Fma:
+    case Opcode::FpTruncWithRounding:
+    case Opcode::Fract:
+    case Opcode::GetBufferDescLength:
+    case Opcode::GetFmaskDescPtr:
+    case Opcode::GetImageDescPtr:
+    case Opcode::GetSamplerDescPtr:
+    case Opcode::GetTexelBufferDescPtr:
+    case Opcode::IndexDescPtr:
+    case Opcode::InsertBitField:
+    case Opcode::IsInf:
+    case Opcode::IsNaN:
+    case Opcode::Ldexp:
+    case Opcode::MatrixTimesMatrix:
+    case Opcode::MatrixTimesScalar:
+    case Opcode::MatrixTimesVector:
+    case Opcode::NormalizeVector:
+    case Opcode::OuterProduct:
+    case Opcode::QuantizeToFp16:
+    case Opcode::Reflect:
+    case Opcode::Refract:
+    case Opcode::SAbs:
+    case Opcode::SMod:
+    case Opcode::SSign:
+    case Opcode::SmoothStep:
+    case Opcode::TransposeMatrix:
+    case Opcode::VectorTimesMatrix:
+    case Power:
+    case Sinh:
+    case Tan:
+    case Tanh:
+      // Functions that don't access memory.
+      func->addFnAttr(Attribute::ReadNone);
+      break;
+    case Opcode::ImageGather:
+    case Opcode::ImageLoad:
+    case Opcode::ImageLoadWithFmask:
+    case Opcode::ImageSample:
+    case Opcode::LoadDescFromPtr:
+    case Opcode::LoadPushConstantsPtr:
+    case Opcode::ReadBuiltInInput:
+    case Opcode::ReadBuiltInOutput:
+    case Opcode::ReadGenericInput:
+    case Opcode::ReadGenericOutput:
+      // Functions that only read memory.
+      func->addFnAttr(Attribute::ReadOnly);
+      break;
+    case Opcode::ImageStore:
+      // Functions that only write memory.
+      func->addFnAttr(Attribute::WriteOnly);
+      break;
+    case Opcode::ImageAtomic:
+    case Opcode::ImageAtomicCompareSwap:
+      // Functions that read and write memory.
+      break;
+    case Opcode::Barrier:
+    case Opcode::DemoteToHelperInvocation:
+    case Opcode::EmitVertex:
+    case Opcode::EndPrimitive:
+    case Opcode::GetSubgroupSize:
+    case Opcode::ImageGetLod:
+    case Opcode::ImageQueryLevels:
+    case Opcode::ImageQuerySamples:
+    case Opcode::ImageQuerySize:
+    case Opcode::IsHelperInvocation:
+    case Opcode::Kill:
+    case Opcode::LoadBufferDesc:
+    case Opcode::ReadClock:
+    case Opcode::SubgroupAll:
```

Comment:
I think all subgroup operations could be marked as readnone since they are at most DPP operations.

Reply:
I tried this in the original commit 947ee51 but it caused 588 failures in dEQP-VK.subgroups.*. So I removed it.

Reply:
At the very least, you'd also have to mark them as `convergent`. Plus, there's the question of the inline asm-hackery. I really need to upstream my control flow changes to LLVM...

Reply:
#770